db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -3957,12 +3957,12 @@ void DBImpl::GetSnapshotContext(
   *snapshot_seqs = snapshots_.GetAll(earliest_write_conflict_snapshot);
 }
 
-Status DBImpl::WaitForCompact(bool wait_unscheduled) {
-  // Wait until the compaction completes
+Status DBImpl::WaitForCompact() {
   InstrumentedMutexLock l(&mutex_);
   while ((bg_bottom_compaction_scheduled_ || bg_compaction_scheduled_ ||
           bg_flush_scheduled_ ||
-          (wait_unscheduled && unscheduled_compactions_)) &&
+          (!shutting_down_.load(std::memory_order_acquire) &&
+           (unscheduled_compactions_ || unscheduled_flushes_))) &&
```

Comment:
If `PauseBackground()` was called, could this function wait forever when there are unscheduled_compactions_?

Reply:
Yes, looks like it. 

But even without this PR, I think this function could still wait forever if there were actually some scheduled jobs at the time of calling `PauseBackgroundWork()`. `(bg_bottom_compaction_scheduled_ > 0 || bg_compaction_scheduled_ > 0 || bg_flush_scheduled_ > 0)`. These three would never go back to 0 unless someone calls `ContinueBackgroundWork()`.

Reply:
To reduce the likelihood of deadlock, should the default behavior be not waiting for unscheduled_* (abort_on_pause = true by default)? Returning Status::Aborted should help prevent users from misusing the API when PauseBackgroundWork() was called.

Reply:
There's a strong case to go that direction (deadlock = bad), but I think I lean slightly in the direction of wait by default, because
(a) there's also a risk that people don't bother checking the return code. This suggests that the default behavior should be to wait until the desired postcondition is satisfied, assuming there's a reasonable expectation that it will be satisfied.
(b) in our world, deadlocks are far from the scariest of failures. Generally they are isolated incidents that don't cause data corruption. A responsible service owner has to deal with unresponsive nodes anyway. If a DB is about to be shipped to many hosts, failing to clear compaction debt first could be much worse, though this is an atypical case.
(c) there is a remote possibility of confusion if someone calls PauseBackgroundWork()+ContinueBackgroundWork() and WaitForCompact() keeps waiting through it, because it never woke up in time to see the pause. Someone might expect it to always Abort.

If we do go with abort on pause by default, a better option name would be wait_through_pause, on the principle of "boolean options are false by default"

Reply:
Thanks for the detail reasoning. I'm fine with either direction as long as these special cases (shutdown/pause) are documented in the API contract.

Reply:
I ended up introducing `abort_on_pause` with default `false`. If someone calls `PauseBackgroundWork()` prior to this, `abort_on_pause` is `false`, we will call `ContinueBackgroundWork()` and wait. If `abort_on_pause` is `true` and the bg jobs were paused, `Status::Aborted` will be returned.

These cases are added to the comment.